naga/src/proc/constant_evaluator.rs

```diff
@@ -592,186 +771,82 @@ impl<'a> ConstantEvaluator<'a> {
-                            ..
-                        },
-                        ..
-                    }
-                ) =>
-            {
-                let mut components: Vec<_> = crate::proc::flatten_compose(
-                    ty0,
-                    src_components0,
-                    self.expressions,
-                    self.types,
+            crate::MathFunction::Asin => {
+                component_wise_float!(self, span, [arg], |e| { e.asin().into() })
+            }
+            crate::MathFunction::Asinh => {
+                component_wise_float!(self, span, [arg], |e| { e.asinh().into() })
+            }
+            crate::MathFunction::Atan => {
+                component_wise_float!(self, span, [arg], |e| { e.atan().into() })
+            }
+            crate::MathFunction::Atanh => {
+                component_wise_float!(self, span, [arg], |e| { e.atanh().into() })
+            }
+            crate::MathFunction::Pow => {
+                component_wise_float!(self, span, [arg, arg1.unwrap()], |e1, e2| {
+                    e1.powf(e2).into()
+                })
+            }
+            crate::MathFunction::Clamp => {
+                component_wise_scalar!(
+                    self,
+                    span,
+                    [arg, arg1.unwrap(), arg2.unwrap()],
+                    |e, low, high| { e.clamp(low, high).into() }
                 )
-                .chain(crate::proc::flatten_compose(
-                    ty1,
-                    src_components1,
-                    self.expressions,
-                    self.types,
-                ))
-                .chain(crate::proc::flatten_compose(
-                    ty2,
-                    src_components2,
-                    self.expressions,
-                    self.types,
-                ))
-                .collect();
-
-                let chunk_size = components.len() / 3;
-                let (es, rem) = components.split_at_mut(chunk_size);
-                let (lows, highs) = rem.split_at(chunk_size);
-                for ((e, low), high) in es.iter_mut().zip(lows).zip(highs) {
-                    *e = self.math_clamp(*e, *low, *high, span)?;
-                }
-                components.truncate(chunk_size);
-
-                Expression::Compose {
-                    ty: ty0,
-                    components,
-                }
             }
-            _ => return Err(ConstantEvaluatorError::InvalidMathArg),
-        };
-
-        self.register_evaluated_expr(expr, span)
+            crate::MathFunction::Cos => {
+                component_wise_float!(self, span, [arg], |e| { e.cos().into() })
+            }
+            crate::MathFunction::Cosh => {
+                component_wise_float!(self, span, [arg], |e| { e.cosh().into() })
+            }
+            crate::MathFunction::Round => {
+                component_wise_float!(self, span, [arg], |e| { e.round().into() })
```

Comment:
This definitely seems like a valid issue. Given that the soonest that `f{32,64}::round_ties_even` can land is 2024-02-02, I see several options to resolve here (all of which should IMO, be punted to another PR):

* Don't even implement this until Rust 1.77.0 (or perhaps later) stabilizes the method. This seems unnecessary, when compared to other options.
* Use an implementation like [`ndarray-ndimage`'s](https://github.com/imeka/ndarray-ndimage/blob/8b14b4d6ecfbc96a8a052f802e342a7049c68d8f/src/lib.rs#L98), which is linked in [the stabilization report for `round_ties_even`](https://github.com/rust-lang/rust/issues/96710#issuecomment-1875226346); for convenience:

	```rust
	// TODO Use x.round_ties_even() when available on stable
	// https://github.com/rust-lang/rust/issues/96710
	fn round_ties_even(x: f64) -> f64 {
	    let i = x as i32;
	    let f = (x - i as f64).abs();
	    if f == 0.5 {
	        if i & 1 == 1 {
	            // -1.5, 1.5, 3.5, ...
	            (x.abs() + 0.5).copysign(x)
	        } else {
	            (x.abs() - 0.5).copysign(x)
	        }
	    } else {
	        x.round()
	    }
	}
	```
* Do one of the above after we land this. I'd prefer to avoid this, since having incorrect behavior for a brief time can be confusing, but maybe it's acceptable?

Reply:
Using a polyfill for now is what I had in mind. I think we can just copy-paste `ndarray-ndimage`'s impl with a permalink where we got it from and a TODO to switch it out once it becomes stable and our MSRV reaches that version.